presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantSortColumns.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.airlift.log.Logger;
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.spi.plan.LogicalProperties;
+import com.facebook.presto.spi.plan.OrderingScheme;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.planner.iterative.GroupReference;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.SortNode;
+
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.sql.planner.plan.Patterns.sort;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+/**
+ * Removes sort columns from input if the source has a Key that refers to the ordering columns
+ */
+public class RemoveRedundantSortColumns
+        implements Rule<SortNode>
+{
+    private static final Logger log = Logger.get(RemoveRedundantSortColumns.class);
+    private static final Pattern<SortNode> PATTERN = sort();
+
+    @Override
+    public Pattern<SortNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(SortNode node, Captures captures, Context context)
+    {
+        OrderingScheme orderingScheme = node.getOrderingScheme();
+        PlanNode source = node.getSource();
+        log.debug("[%s] SortNode : %s", node.getId(), node);
+
+        Optional<LogicalProperties> sourceLogicalProperties = ((GroupReference) source).getLogicalProperties();
```

Comment:
Is there any case where `node.getSource()` is not an instance of `GroupReference`?

Reply:
Also, why not just use `SortNode#computeLogicalProperties` and pass a `LogicalPropertiesProvider`?

Reply:
No, in the IterativeOptiimizer `getSource()` will always return a GroupReference.  
We precompute the logical properties via `computeLogicalProperities`  when creating a group, so we don't need to redo the compute here

https://github.com/prestodb/presto/blob/e5625be113d7a45b64d5f3027bee86949979a227/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Memo.java#L265-L266

See [these rules](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java#L544-L554) as a reference